### ISSUE TITLE
SharedMemory.close() destroys memory

### ISSUE THREAD

**Issue by 9a6edf97-614f-4265-93ba-c124aa80e594** [python/cpython#91044]
BPO | [46888](https://bugs.python.org/issue46888)
--- | :---
Nosy | @pfmoore, @tjguk, @zware, @eryksun, @zooba, @ronny-rentner

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-03-01.11:39:08.468>
labels = ['type-bug', '3.9', '3.10', '3.11', 'OS-windows', 'docs']
title = 'SharedMemory.close() destroys memory'
updated_at = <Date 2022-03-02.11:33:21.625>
user = 'https://github.com/ronny-rentner'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-02.11:33:21.625>
actor = 'eryksun'
assignee = 'docs@python'
closed = False
closed_date = None
closer = None
components = ['Documentation', 'Windows']
creation = <Date 2022-03-01.11:39:08.468>
creator = 'ronny-rentner'
dependencies = []
files = []
hgrepos = []
issue_num = 46888
keywords = []
message_count = 8.0
messages = ['414258', '414259', '414260', '414324', '414331', '414335', '414338', '414339']
nosy_count = 7.0
nosy_names = ['paul.moore', 'tim.golden', 'docs@python', 'zach.ware', 'eryksun', 'steve.dower', 'ronny-rentner']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue46888'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

**Comment by jamesra** [comment]
I've found this frustrating bug after putting the work into moving to SharedMemory and not seeing this Windows issue mentioned in the documentation (please add that).  A workable fix for Windows users would be that if a SharedMemoryManager process is used to create the SharedMemory the manager process keeps SharedMemory instances alive until unlink is called.  This does not appear to happen now.   My expectation is even if the child processes close the shared memory block before the parent process can open the shared memory the block is still be available until I call unlink or the shared memory manager process closes. 

Shared memory is much faster in my testing compared to my old method of passing filename strings and child processes pulling the data from disk.  I'd really like to use it to return arbitrary sized results to the parent process.

**Comment by jsdpag** [comment]
> I've found this frustrating bug after putting the work into moving to SharedMemory and not seeing this Windows issue mentioned in the documentation (please add that). A workable fix for Windows users would be that if a SharedMemoryManager process is used to create the SharedMemory the manager process keeps SharedMemory instances alive until unlink is called. This does not appear to happen now. My expectation is even if the child processes close the shared memory block before the parent process can open the shared memory the block is still be available until I call unlink or the shared memory manager process closes.
> 
> Shared memory is much faster in my testing compared to my old method of passing filename strings and child processes pulling the data from disk. I'd really like to use it to return arbitrary sized results to the parent process.

Agreed. It would be nice if this inconsistency in behaviour was mentioned in the documentation.
